lib/src/native/realm_core.dart

```diff
@@ -279,24 +286,18 @@ class _RealmCore {
     return result == nullptr ? null : SubscriptionHandle._(result);
   }
 
-  static void _stateChangeCallback(Pointer<Void> userdata, int state) {
-    final completer = userdata.toObject<Completer<SubscriptionSetState>>(isPersistent: true);
-    if (completer == null) {
-      return;
-    }
+  static void _stateChangeCallback(Object userdata, int state) {
+    final completer = userdata as Completer<SubscriptionSetState>;
+
     completer.complete(SubscriptionSetState.values[state]);
   }
 
   Future<SubscriptionSetState> waitForSubscriptionSetStateChange(SubscriptionSet subscriptions, SubscriptionSetState notifyWhen) {
     final completer = Completer<SubscriptionSetState>();
-    _realmLib.realm_dart_sync_on_subscription_set_state_change_async(
-      subscriptions.handle._pointer,
-      notifyWhen.index,
-      Pointer.fromFunction(_stateChangeCallback),
-      completer.toPersistentHandle(),
-      _realmLib.addresses.realm_dart_delete_persistent_handle,
-      scheduler.handle._pointer,
-    );
+    final callback = Pointer.fromFunction<Void Function(Handle, Int32)>(_stateChangeCallback);
+    final userdata = _realmLib.realm_dart_userdata_async_new(completer, callback.cast(), scheduler.handle._pointer);
```

Comment:
We could introduce a extension method, so that we could write something like:
```dart
final userdata = completer.asUserdataAsync(callback.cast());
```

Reply:
I really think we should try allocating this struct in Dart.

Reply:
ignore we were doing two calls before